### ISSUE TITLE
Generic OAuth: Tries to access $API_URL/emails and fails if HTML is returned

### ISSUE THREAD

**Issue by waldiTM** [grafana/grafana#68678]
### 
# What went wrong?

**What happened**:

Generic OAuth2 with GitLab as backend for OIDC fails with:
```json
{
  "error":"invalid character '<' looking for beginning of value",
  "level":"error",
  "logger":"oauth.generic_oauth",
  "msg":"Error decoding email addresses response"
}
```

It did not request scope "email" by mistake, so that information was missing.

To validate the user, Grafana tries to access:
- $TOKEN_URL (works),
- $API_URL (aka userinfo, works, but does not return an email field) 
- $API_URL/emails (redirects to the sign in page and then it tries to parse html)

The complete HTML is dumped into the log as well.

**What did you expect to happen**:

Login in works. It does not break if email is missing.

### 
# How do we reproduce it?

**Step 1**:

- Open Grafana and do OAuth

**Step 2**:

- Wait for the redirect, it will juust say: `login.OAuthLogin(get info from generic_oauth)`

### 
# What Grafana version are you using?

v9.5.2, aka docker.io/grafana/grafana-oss@sha256:2aafd24a138277142a86b6ed43b7537e1ec8226478f5240c459e9330a3461cdb

### 
### Grafana Platform?

Docker

### User's OS?

Debian 12

### Is this a Regression?

None

**Comment by mellieA** [comment]
@waldiTM the expectation is that oauth will search for the email and then set it to a blank string if it is missing: https://grafana.com/docs/grafana/latest/setup-grafana/configure-security/configure-authentication/generic-oauth/#email-address

The line of code that returns the error you saw means that there was an error decoding the json response that is expected to hold the email address if you're using JMES for role mapping. Was there something returned in the response that was interpreted as a malformed email address?

That section of code doesn't look like it's been changed recently. Are you able to share a sanitized snippet of the config files you used when you saw this error?

**Comment by waldiTM** [comment]
@mellieA Exactly. This URL is not supported on this IdP. So Grafana get's a HTTP redirect and then a HTML page back. There is no JSON involved that could be decoded. And Grafana clearly should not log several KiB of HTML it can't ever read.

I also found no mention of this endpoint in the [OIDC spec](https://openid.net/specs/openid-connect-core-1_0.html#UserInfo), so I assume this is something special to a particular implementation.

Why I saw this error is easy to describe: the oauth request was missing one scope, so no `email` attribute was provided.

**Comment by UKFr-DIZ** [comment]
@mellieA we have a similar issue still. 
We use keycloak `Version 23.0.6` as an IdP and we also couldn't find this particular endpoint `https://<kc-url>/auth/realms/<realm>/protocol/openid-connect/userinfo/emails` OIDC spec. The IdP return a 404 error which causes the login to fail. 

Error message
~~~
logger=oauth.generic_oauth t=2024-01-17T14:06:24.117408923Z level=error msg="Error getting email address" url=https://<kc-url>/auth/realms/<realm>/protocol/openid-connect/userinfo/emails error="{\"error\":\"RESTEASY003210: Could not find resource for full path: https://<kc-url>/auth/realms/<realm>/protocol/openid-connect/userinfo/emails\"}"
logger=context userId=0 orgId=0 uname= t=2024-01-17T14:06:24.117458455Z level=error msg="login.OAuthLogin(get info from generic_oauth)" error="Error getting email address: {\"error\":\"RESTEASY003210: Could not find resource for full path: https://<kc-url>/auth/realms/<realm>/protocol/openid-connect/userinfo/emails\"}"
logger=context userId=0 orgId=0 uname= t=2024-01-17T14:06:24.117692252Z level=error msg="Request Completed" method=GET path=/login/generic_oauth status=500 remote_addr=10.170.24.251 time_ms=261 duration=261.506869ms size=1372 referer= handler=/login/:name
~~~

Checking for an addtional endpoint is also not necessary, given that the userinfo endpoint return `email ` attribute already in the firstplace. 

by removing the few lines only https://github.com/grafana/grafana/blob/9bbb7f67e0d2bb884d40d0064de0dfdb5c2f3966/pkg/login/social/connectors/generic_oauth.go#L477

this issue could be solved i think.

**Comment by fredleb** [comment]
I'm having exactly the same problem. Was that ever solved or is there a workaround or something to setup in keycloak ?

**Comment by Julien-Delavisse** [comment]
I don't know if it's the same problem I had, but adding the `openid` scope solved mine.

**Comment by T-baby** [comment]
You can try removing the default scopes and adding `openid email` to your openid scope; this resolved the issue for me.
